### ISSUE TITLE
Add transcoding for video

### ISSUE THREAD

**Issue by derneuere** [LibrePhotos/librephotos#365]
As a user, I want to view video files that are not native to the browser. 

I looked for that in the implementation in [stash](https://github.com/stashapp/stash/blob/6ce57a9a43e876f7f6dddcfb38ad0078a7d1defa/pkg/api/routes_scene.go#L34). Sadly it is implemented in Go, but the general structure will be useful for our implementation:

- Add a new API endpoint for requesting a transcode for video with a starting point. Should close the process when not playing the video.
- Add two functions for creating a new mp4 file and for transcoding the video via ffmpeg
- Add the mp4 option in [LightBox](https://github.com/LibrePhotos/librephotos-frontend/blob/6f4132899d99e552a1bf4b847c3c5ce0b7117438/src/components/lightbox/LightBox.js#L15), when it's seeking and transcoding, then start a new transcoding process via the API and activate it when the requested file is not in the right format.

**Comment by derneuere** [comment]
Here is a short article about that: https://cjwebb.com/python-docker-ffmpeg-h264-mp4/

Here is an example on how to do something similiar in python and flask: https://gist.github.com/anthonyeden/f3b3bdf6f62badd8f87bb574283f488a

We should then use django streaming response: https://stackoverflow.com/questions/38514919/django-stream-request-from-external-site-as-received

**Comment by derneuere** [comment]
An experimental option was added. StreamingHTTPResponse does not seem to stop once you close the light box, but it does when the window is closed. I have to investigate what possible fixes are.

We can use the same endpoint, but add a bit of information like "?s=109&r=1080" or something like that. This will obviously only be relevant when transcoding.

We also have to write a custom ReactPlayer to override the duration, as the video file from ffmpeg does not know while transcoding how long it will be. This should be added to the photo model as an attribute.

**Comment by derneuere** [comment]
Now the process closes reliably.

**Comment by derneuere** [comment]
[flaurafauna](https://github.com/flaurafauna) noted that it currently uses all available resources when transcoding. This should be more limited to be a bit faster than real-time, as more transcoding power is not needed for viewing.

Also, sometimes the pause / scrub bars missing. Not sure if that's also related to transcoding.

**Comment by derneuere** [comment]
Closing — transcoding shipped.

`VideoTranscoder` pipes the original through ffmpeg (libx264, `frag_keyframe+empty_moov`, scaled to 720) and `_transcoded_video_response` returns a `StreamingHttpResponse` of `video/mp4` — `apps/backend/api/views/views.py:533-611`. It is gated on the per-user `transcode_videos` field (`api/models/user.py:58`, migration `0016_add_transcode_videos`), exposed in the settings UI as `settings.transcodevideo`, and consumed by the lightbox via `${serverAddress}/media/photos/${mediaHash}.mp4` (`apps/frontend/src/components/lightbox/MediaDisplay.tsx:71-85`).

The last blocker in the original plan was closed out by the 2022-05-19 comment in this thread ("Now the process closes reliably").

**Still open, tracked separately:** the resource-usage point raised at the end of this thread — ffmpeg is spawned with no thread cap and no rate limit — is now #1920.
